pkg/autodiscovery/providers/kube_endpoints_file.go

```diff
@@ -0,0 +1,310 @@
+	s.Lock()
+	defer s.Unlock()
+
+	_, found := s.epConfigs[id]
+	if !found {
+		s.epConfigs[id] = newEpConfig()
+	}
+
+	s.epConfigs[id].templates = append(s.epConfigs[id].templates, tpl)
+}
+
+// insertEp caches the provided endpoints object if it matches one of the tracked configs
+// and prepares the config to be collected in the next Collect call.
+// Returns false if the endpoints object is irrelevant and discarded.
+func (s *store) insertEp(ep *v1.Endpoints) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	epConfig, found := s.epConfigs[epID(ep.Namespace, ep.Name)]
+	if !found {
+		return false
+	}
+
+	epConfig.ep = ep
+	epConfig.shouldCollect = true
+
+	return true
+}
+
+// deleteEp handles endpoint objects deletion.
+// it marks the corresponfing config as not collectable.
+func (s *store) deleteEp(ep *v1.Endpoints) {
+	s.Lock()
+	defer s.Unlock()
+
+	epConfig, found := s.epConfigs[epID(ep.Namespace, ep.Name)]
+	if !found {
+		return
+	}
+
+	epConfig.ep = nil
+	epConfig.shouldCollect = false
+}
+
+func (s *store) isEmpty() bool {
+	s.RLock()
+	defer s.RUnlock()
+
+	return len(s.epConfigs) == 0
+}
+
+// generateConfigs transforms the cached config templates into collectable integration.Config objects
+func (s *store) generateConfigs() []integration.Config {
+	s.Lock()
+	defer s.Unlock()
+
+	configs := []integration.Config{}
+	for _, epConfig := range s.epConfigs {
+		if epConfig.shouldCollect {
+			for _, tpl := range epConfig.templates {
+				configs = append(configs, endpointChecksFromTemplate(tpl, epConfig.ep)...)
+			}
+		}
+	}
+
+	return configs
+}
+
+// endpointChecksFromTemplate resolves an integration.Config template based on the provided Endpoints object.
+func endpointChecksFromTemplate(tpl integration.Config, ep *v1.Endpoints) []integration.Config {
+	configs := []integration.Config{}
+	if ep == nil {
+		return configs
+	}
+
+	for i := range ep.Subsets {
+		for j := range ep.Subsets[i].Addresses {
+			entity := apiserver.EntityForEndpoints(ep.Namespace, ep.Name, ep.Subsets[i].Addresses[j].IP)
+			config := &integration.Config{
+				Entity:                  entity,
+				Name:                    tpl.Name,
+				Instances:               tpl.Instances,
+				InitConfig:              tpl.InitConfig,
+				MetricConfig:            tpl.MetricConfig,
+				LogsConfig:              tpl.LogsConfig,
+				ADIdentifiers:           []string{entity},
+				AdvancedADIdentifiers:   nil,
+				ClusterCheck:            true,
+				Provider:                names.KubeEndpointsFile,
+				Source:                  tpl.Source,
+				IgnoreAutodiscoveryTags: tpl.IgnoreAutodiscoveryTags,
+			}
+
+			utils.ResolveEndpointConfigAuto(config, ep.Subsets[i].Addresses[j])
+			configs = append(configs, *config)
+		}
+	}
+
+	return configs
+}
+
+func epID(ns, name string) string {
+	return ns + "/" + name
+}
+
+func init() {
+	RegisterProvider(names.KubeEndpointsFileRegisterName, NewKubeEndpointsFileConfigProvider)
+}
```

Comment:
could you move the `init()` at the top of the file?

Reply:
done